File: two_sum/two_sum.go
package main

import (
	"errors"
	"fmt"
)

func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
}

func twoSum(nums []int, target int) ([]int, error) {
	prevNums := make(map[int]int)

	for idx, num := range nums {
		diff := target - num

		if requiredIdx, ok := prevNums[diff]; ok {
			return []int{requiredIdx, idx}, nil
		}

		prevNums[num] = idx
	}

	return []int{}, errors.New("not found")
}

File: two_sum/twoSum.js
/**
 * @param {number[]} nums - The array of numbers.
 * @returns {number[]} The two indexes of the values that add to target
 */
function twoSum(nums, target) {
  // number is key, index is value
  const map = {};

  for (let i = 0; i < nums.length; i++) {
    const num = nums[i];
    const diff = target - num;

    if (diff in map) {
      return [map[diff], i];
    }

    map[num] = i;
  }
}
